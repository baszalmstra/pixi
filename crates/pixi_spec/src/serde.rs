use rattler_conda_types::{
    BuildNumberSpec, NamedChannelOrUrl, ParseStrictness::Strict, StringMatcher, VersionSpec,
};
use rattler_digest::{Md5Hash, Sha256Hash};
use serde::{de::Error, Deserialize, Deserializer, Serialize, Serializer};
use serde_with::serde_as;
use url::Url;

use crate::{DetailedVersionSpec, GitRev, GitSpec, PathSpec, Spec, UrlSpec};

#[serde_as]
#[derive(serde::Deserialize)]
#[serde(deny_unknown_fields)]
struct RawSpec {
    /// The version spec of the package (e.g. `1.2.3`, `>=1.2.3`, `1.2.*`)
    #[serde_as(as = "Option<serde_with::DisplayFromStr>")]
    pub version: Option<VersionSpec>,

    /// The URL of the package
    pub url: Option<Url>,

    /// The git url of the package
    pub git: Option<Url>,

    /// The path to the package
    pub path: Option<String>,

    /// The git revision of the package
    pub branch: Option<String>,

    /// The git revision of the package
    pub rev: Option<String>,

    /// The git revision of the package
    pub tag: Option<String>,

    /// The build string of the package (e.g. `py37_0`, `py37h6de7cb9_0`, `py*`)
    #[serde_as(as = "Option<serde_with::DisplayFromStr>")]
    pub build: Option<StringMatcher>,

    /// The build number of the package
    pub build_number: Option<BuildNumberSpec>,

    /// Match the specific filename of the package
    pub file_name: Option<String>,

    /// The channel of the package
    pub channel: Option<NamedChannelOrUrl>,

    /// The subdir of the channel
    pub subdir: Option<String>,

    /// The md5 hash of the package
    #[serde_as(as = "Option<rattler_digest::serde::SerializableHash::<rattler_digest::Md5>>")]
    pub md5: Option<Md5Hash>,

    /// The sha256 hash of the package
    #[serde_as(as = "Option<rattler_digest::serde::SerializableHash::<rattler_digest::Sha256>>")]
    pub sha256: Option<Sha256Hash>,
}

impl<'de> Deserialize<'de> for Spec {
    fn deserialize<D>(deserializer: D) -> Result<Spec, D::Error>
    where
        D: Deserializer<'de>,
    {
        serde_untagged::UntaggedEnumVisitor::new()
            .expecting(
                "a version string like \">=0.9.8\" or a detailed dependency like { version = \">=0.9.8\" }",
            )
            .string(|str| {
                VersionSpec::from_str(str, Strict)
                    .map_err(serde_untagged::de::Error::custom)
                    .map(Spec::Version)
            })
            .map(|map| {
                let raw_spec: RawSpec = map.deserialize()?;

                if raw_spec.git.is_none()
                    && (raw_spec.branch.is_some()
                        || raw_spec.rev.is_some()
                        || raw_spec.tag.is_some())
                {
                    return Err(serde_untagged::de::Error::custom(
                        "`branch`, `rev`, and `tag` are only valid when `git` is specified",
                    ));
                }

                if raw_spec.version.is_none() && raw_spec.build.is_some() {
                    return Err(serde_untagged::de::Error::custom(
                        "`build` is only valid when `version` is specified",
                    ));
                }

                if raw_spec.version.is_none() && raw_spec.build_number.is_some() {
                    return Err(serde_untagged::de::Error::custom(
                        "`build-number` is only valid when `version` is specified",
                    ));
                }

                if raw_spec.version.is_none() && raw_spec.file_name.is_some() {
                    return Err(serde_untagged::de::Error::custom(
                        "`file-name` is only valid when `version` is specified",
                    ));
                }

                if raw_spec.version.is_none() && raw_spec.channel.is_some() {
                    return Err(serde_untagged::de::Error::custom(
                        "`channel` is only valid when `version` is specified",
                    ));
                }

                if raw_spec.version.is_none() && raw_spec.subdir.is_some() {
                    return Err(serde_untagged::de::Error::custom(
                        "`subdir` is only valid when `version` is specified",
                    ));
                }

                if raw_spec.version.is_none() && raw_spec.url.is_none() && raw_spec.sha256.is_some()
                {
                    return Err(serde_untagged::de::Error::custom(
                        "`sha256` is only valid when `version` or `url` is specified",
                    ));
                }

                if raw_spec.version.is_none() && raw_spec.url.is_none() && raw_spec.md5.is_some() {
                    return Err(serde_untagged::de::Error::custom(
                        "`md5` is only valid when `version` or `url` is specified",
                    ));
                }

                let spec = match (raw_spec.version, raw_spec.url, raw_spec.path, raw_spec.git) {
                    (Some(version), None, None, None) => Spec::DetailedVersion(DetailedVersionSpec {
                        version,
                        build: raw_spec.build,
                        build_number: raw_spec.build_number,
                        file_name: raw_spec.file_name,
                        channel: raw_spec.channel,
                        subdir: raw_spec.subdir,
                        md5: raw_spec.md5,
                        sha256: raw_spec.sha256,
                    }),
                    (None, Some(url), None, None) => Spec::Url(UrlSpec {
                        url,
                        md5: raw_spec.md5,
                        sha256: raw_spec.sha256,
                    }),
                    (None, None, Some(path), None) => Spec::Path(PathSpec { path: path.into() }),
                    (None, None, None, Some(git)) => {
                        let rev = match (raw_spec.branch, raw_spec.rev, raw_spec.tag) {
                            (Some(branch), None, None) => Some(GitRev::Branch(branch)),
                            (None, Some(rev), None) => Some(GitRev::Commit(rev)),
                            (None, None, Some(tag)) => Some(GitRev::Tag(tag)),
                            (None, None, None) => None,
                            _ => {
                                return Err(serde_untagged::de::Error::custom(
                                    "only one of `branch`, `rev`, or `tag` can be specified",
                                ));
                            }
                        };
                        Spec::Git(GitSpec { git, rev })
                    }
                    (None, None, None, None) => {
                        return Err(serde_untagged::de::Error::custom(
                            "one of `version`, `url`, or `path` must be specified",
                        ))
                    }
                    (_, _, _, _) => {
                        return Err(serde_untagged::de::Error::custom(
                            "only one of `version`, `url`, `path`, or `git` can be specified",
                        ))
                    }
                };

                Ok(spec)
            })
            .deserialize(deserializer)
    }
}

impl<'de> Deserialize<'de> for PathSpec {
    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
    where
        D: Deserializer<'de>,
    {
        #[derive(Deserialize)]
        struct Raw {
            path: String,
        }

        Raw::deserialize(deserializer).map(|raw| PathSpec {
            path: raw.path.into(),
        })
    }
}

impl Serialize for PathSpec {
    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: Serializer,
    {
        #[derive(Serialize)]
        struct Raw {
            path: String,
        }

        Raw {
            path: self.path.to_string(),
        }
        .serialize(serializer)
    }
}

#[cfg(test)]
mod test {
    use serde::Serialize;
    use serde_json::{json, Value};

    use super::*;

    #[test]
    fn test_round_trip() {
        let examples = [
            json! { "1.2.3" },
            json!({ "version": "1.2.3" }),
            json!({ "ver": "1.2.3" }),
            json! { "*" },
            json!({ "path": "foobar" }),
            json!({ "path": "foobar", "version": "1.2.3" }),
            json!({ "version": "//" }),
            json!({ "path": "foobar", "version": "//" }),
            json!({ "path": "foobar", "sha256": "315f5bdb76d078c43b8ac0064e4a0164612b1fce77c869345bfc94c75894edd3" }),
            json!({ "version": "1.2.3", "sha256": "315f5bdb76d078c43b8ac0064e4a0164612b1fce77c869345bfc94c75894edd3" }),
            json!({ "url": "https://conda.anaconda.org/conda-forge/linux-64/21cmfast-3.3.1-py38h0db86a8_1.conda" }),
            json!({ "url": "https://conda.anaconda.org/conda-forge/linux-64/21cmfast-3.3.1-py38h0db86a8_1.conda", "sha256": "315f5bdb76d078c43b8ac0064e4a0164612b1fce77c869345bfc94c75894edd3" }),
            json!({ "git": "https://github.com/conda-forge/21cmfast-feedstock" }),
            json!({ "git": "https://github.com/conda-forge/21cmfast-feedstock", "branch": "main" }),
            json!({ "git": "https://github.com/conda-forge/21cmfast-feedstock", "branch": "main", "tag": "v1" }),
        ];

        #[derive(Serialize)]
        struct Snapshot {
            input: Value,
            result: Value,
        }

        let mut snapshot = Vec::new();
        for input in examples {
            let spec: Result<Spec, _> = serde_json::from_value(input.clone());
            let result = match spec {
                Ok(spec) => {
                    let spec = Spec::from(spec);
                    serde_json::to_value(&spec).unwrap()
                }
                Err(e) => {
                    json!({
                        "error": format!("ERROR: {e}")
                    })
                }
            };

            snapshot.push(Snapshot { input, result });
        }

        insta::assert_yaml_snapshot!(snapshot);
    }
}
