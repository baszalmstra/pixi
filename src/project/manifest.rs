use crate::command::Command;
use ::serde::Deserialize;
use rattler_conda_types::{Channel, Platform, Version};
use rattler_virtual_packages::{Archspec, Cuda, LibC, Linux, Osx, VirtualPackage};
use serde_with::{serde_as, DisplayFromStr};
use std::collections::HashMap;

/// Describes the contents of a project manifest.
#[derive(Debug, Clone, Deserialize)]
pub struct ProjectManifest {
    /// Information about the project
    pub project: ProjectMetadata,

    /// Commands defined in the project
    #[serde(default)]
    pub commands: HashMap<String, Command>,

    /// Additional system requirements
    #[serde(default, rename = "system-requirements")]
    pub system_requirements: SystemRequirements,
}

/// Describes the contents of the `[package]` section of the project manifest.
#[serde_as]
#[derive(Debug, Clone, Deserialize)]
pub struct ProjectMetadata {
    /// The name of the project
    pub name: String,

    /// The version of the project
    #[serde_as(as = "DisplayFromStr")]
    pub version: Version,

    /// An optional project description
    pub description: Option<String>,

    /// Optional authors
    #[serde(default)]
    pub authors: Vec<String>,

    /// The channels used by the project
    #[serde_as(deserialize_as = "Vec<super::serde::ChannelStr>")]
    pub channels: Vec<Channel>,

    /// The platforms this project supports
    // TODO: This is actually slightly different from the rattler_conda_types::Platform because it
    //     should not include noarch.
    pub platforms: Vec<Platform>,
}

#[serde_as]
#[derive(Debug, Clone, Deserialize, Default)]
#[serde(deny_unknown_fields)]
pub struct SystemRequirements {
    windows: Option<bool>,

    unix: Option<bool>,

    #[serde_as(as = "Option<DisplayFromStr>")]
    macos: Option<Version>,

    #[serde_as(as = "Option<DisplayFromStr>")]
    linux: Option<Version>,

    #[serde_as(as = "Option<DisplayFromStr>")]
    cuda: Option<Version>,

    libc: Option<LibCSystemRequirement>,

    archspec: Option<String>,
}

impl SystemRequirements {
    pub fn virtual_packages(&self) -> Vec<VirtualPackage> {
        let mut result = Vec::new();
        if self.windows == Some(true) {
            result.push(VirtualPackage::Win);
        }
        if self.unix == Some(true) {
            result.push(VirtualPackage::Unix);
        }
        if let Some(version) = self.linux.clone() {
            result.push(VirtualPackage::Linux(Linux { version }));
        }
        if let Some(version) = self.cuda.clone() {
            result.push(VirtualPackage::Cuda(Cuda { version }));
        }
        if let Some(version) = self.macos.clone() {
            result.push(VirtualPackage::Osx(Osx { version }))
        }
        if let Some(spec) = self.archspec.clone() {
            result.push(VirtualPackage::Archspec(Archspec { spec }))
        }
        if let Some(libc) = self.libc.clone() {
            result.push(VirtualPackage::LibC(libc.into()))
        }

        result
    }
}

#[serde_as]
#[derive(Debug, Clone, Deserialize)]
#[serde(untagged)]
pub enum LibCSystemRequirement {
    /// Only a version was specified, we assume glibc.
    GlibC(#[serde_as(as = "DisplayFromStr")] Version),

    /// Specified both a family and a version.
    OtherFamily(LibCFamilyAndVersion),
}

#[serde_as]
#[derive(Debug, Clone, Deserialize)]
#[serde(deny_unknown_fields)]
pub struct LibCFamilyAndVersion {
    /// The libc family, e.g. glibc
    pub family: Option<String>,

    /// The minimum version of the libc family
    #[serde_as(as = "DisplayFromStr")]
    pub version: Version,
}

impl From<LibCSystemRequirement> for LibC {
    fn from(value: LibCSystemRequirement) -> Self {
        match value {
            LibCSystemRequirement::GlibC(version) => LibC {
                version,
                family: String::from("glibc"),
            },
            LibCSystemRequirement::OtherFamily(libc) => libc.into(),
        }
    }
}

impl From<LibCFamilyAndVersion> for LibC {
    fn from(value: LibCFamilyAndVersion) -> Self {
        LibC {
            version: value.version,
            family: value.family.unwrap_or_else(|| String::from("glibc")),
        }
    }
}
