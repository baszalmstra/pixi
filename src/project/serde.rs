use rattler_conda_types::{Channel, ChannelConfig};
use serde::de::Error;
use serde::{Deserialize, Deserializer};
use std::borrow::Cow;

pub struct ChannelStr;

impl<'de> serde_with::DeserializeAs<'de, Channel> for ChannelStr {
    fn deserialize_as<D>(deserializer: D) -> Result<Channel, D::Error>
    where
        D: Deserializer<'de>,
    {
        let channel_str = Cow::<str>::deserialize(deserializer)?;
        let channel_config = ChannelConfig::default();
        Channel::from_str(channel_str, &channel_config).map_err(D::Error::custom)
    }
}
