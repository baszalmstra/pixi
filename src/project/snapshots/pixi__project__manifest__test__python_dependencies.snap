---
source: src/project/manifest.rs
assertion_line: 518
expression: "toml_edit::de::from_str::<ProjectManifest>(&contents).expect(\"parsing should succeed!\")"
---
ProjectManifest {
    project: ProjectMetadata {
        name: "foo",
        version: Some(
            Version {
                version: [[0], [0], [1], [0]],
                local: [],
            },
        ),
        description: None,
        authors: [],
        channels: [],
        platforms: PixiSpanned {
            span: Some(
                121..123,
            ),
            value: [],
        },
        license: None,
        license_file: None,
        readme: None,
        homepage: None,
        repository: None,
        documentation: None,
    },
    tasks: {},
    system_requirements: SystemRequirements {
        windows: None,
        unix: None,
        macos: None,
        linux: None,
        cuda: None,
        libc: None,
        archspec: None,
    },
    dependencies: {},
    host_dependencies: None,
    build_dependencies: None,
    target: {},
    activation: None,
    pypi_dependencies: PypiDependencies {
        requirements: {
            PixiSpanned {
                span: Some(
                    177..180,
                ),
                value: PackageName {
                    source: "foo",
                    normalized: "foo",
                },
            }: PixiSpanned {
                span: Some(
                    183..191,
                ),
                value: (
                    Dependency(
                        VersionSpecifier(
                            VersionSpecifiers(
                                [
                                    VersionSpecifier {
                                        operator: GreaterThanEqual,
                                        version: Version {
                                            epoch: 0,
                                            release: [
                                                3,
                                                12,
                                            ],
                                            pre: None,
                                            post: None,
                                            dev: None,
                                            local: None,
                                        },
                                    },
                                ],
                            ),
                        ),
                    ),
                    Requirement {
                        name: "foo",
                        extras: None,
                        version_or_url: Some(
                            VersionSpecifier(
                                VersionSpecifiers(
                                    [
                                        VersionSpecifier {
                                            operator: GreaterThanEqual,
                                            version: Version {
                                                epoch: 0,
                                                release: [
                                                    3,
                                                    12,
                                                ],
                                                pre: None,
                                                post: None,
                                                dev: None,
                                                local: None,
                                            },
                                        },
                                    ],
                                ),
                            ),
                        ),
                        marker: None,
                    },
                ),
            },
            PixiSpanned {
                span: Some(
                    204..207,
                ),
                value: PackageName {
                    source: "bar",
                    normalized: "bar",
                },
            }: PixiSpanned {
                span: Some(
                    210..246,
                ),
                value: (
                    LongDependency {
                        version: VersionSpecifier(
                            VersionSpecifiers(
                                [
                                    VersionSpecifier {
                                        operator: GreaterThanEqual,
                                        version: Version {
                                            epoch: 0,
                                            release: [
                                                3,
                                                12,
                                            ],
                                            pre: None,
                                            post: None,
                                            dev: None,
                                            local: None,
                                        },
                                    },
                                ],
                            ),
                        ),
                        common: RequirementModernCommon {
                            optional: false,
                            extras: Some(
                                [
                                    "baz",
                                ],
                            ),
                            markers: None,
                        },
                    },
                    Requirement {
                        name: "bar",
                        extras: Some(
                            [
                                "baz",
                            ],
                        ),
                        version_or_url: Some(
                            VersionSpecifier(
                                VersionSpecifiers(
                                    [
                                        VersionSpecifier {
                                            operator: GreaterThanEqual,
                                            version: Version {
                                                epoch: 0,
                                                release: [
                                                    3,
                                                    12,
                                                ],
                                                pre: None,
                                                post: None,
                                                dev: None,
                                                local: None,
                                            },
                                        },
                                    ],
                                ),
                            ),
                        ),
                        marker: None,
                    },
                ),
            },
        },
    },
}
